dbms/src/Storages/DeltaMerge/DeltaMerge.h

```diff
@@ -247,14 +267,36 @@ class DeltaMergeBlockInputStream final : public SkippableBlockInputStream
             if (limit == max_block_size)
                 continue;
 
-            return header.cloneWithColumns(std::move(columns));
+            auto block = header.cloneWithColumns(std::move(columns));
+            if constexpr (need_row_id)
+            {
+                block.setSegmentRowIdCol(std::move(seg_row_id_col));
+            }
+            return block;
         }
         return {};
     }
 
 private:
     inline bool finished() { return stable_done && delta_done; }
 
+    inline void fillSegmentRowId(UInt64 start, UInt64 limit)
+    {
+        for (UInt64 i = 0; i < limit; i++)
+        {
+            UInt32 row_id = start + i;
+            seg_row_id_col->insertData(reinterpret_cast<const char*>(&row_id), sizeof(row_id));
+        }
+    }
+
+    inline void fillSegmentRowId(const std::vector<UInt32> & row_ids)
+    {
+        for (UInt32 row_id : row_ids)
+        {
+            row_id += stable_rows;
+            seg_row_id_col->insertData(reinterpret_cast<const char*>(&row_id), sizeof(row_id));
+        }
+    }
```

Comment:
how about moving the two function to inside `class Block`?

Reply:
Since we need fillSegmentRowId before the filters, move inside `class Block` can help reuse the logic.

Reply:
When implementing late materialization, we need SegmentRowId to generate fitering-bitmap.